ttl: finalize for v22.1 release

otan opened the pull request:
See individual commits for details.

Big release note on the last commit.

otan committed:
tabledesc: ensure no descending order PKs for row level TTL

Supporting descending order PKs is complex as it requires a more
complicated pagination algorithm. As such, let's not support such a
table for now.

Release note: None

otan committed:
CODEOWNERS: add sql-experience to CODEOWNERS for row-level TTL

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76918)
<!-- Reviewable:end -->

otan committed:
tabledesc: ensure crdb_internal_expiration column expressions match

Release note: None

otan committed:
sql: do not allow SET (ttl_expire_after) if automatic column defined

Note we allow this for CREATE TABLE so SHOW CREATE TABLE round trips.
But we should never need this for the ALTER TABLE ... SET ...

Release note: None

otan committed:
sql: ensure SHOW CREATE TABLE matches pg_class.reloptions

Release note: None

rafiss reviewed (approved):
lgtm!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76918#-:-MwlUhjaA-RsMZ5B_elL:b-nzoorr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[-- commits, line 40 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76918#-MwlUE3m5IRICNRci0Hl):*
nit: "which is available as a beta feature"
___
*[-- commits, line 65 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76918#-MwlUV8g2wzvbVmqesqB):*
nit: maybe mention that the rate is per second?
___
*[pkg/sql/catalog/tabledesc/structured.go, line 2527 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76918#-MwlU0TK9dfoRuv7Artf:-MwlU0TK9dfoRuv7Artg:bx70mw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b356c43de90fa473be4c43d1e81e67880afd178/pkg/sql/catalog/tabledesc/structured.go#L2527)):*
> ```Go
> 
> // GetStorageParams implements the TableDescriptor interface.
> func (desc *wrapper) GetStorageParams(spaceBetweenEqual bool) []string {
> ```

i wonder if there's a good way to make sure we don't forget to update this function when new storage params are added


<!-- Sent from Reviewable.io -->

otan committed:
ttl: annotate post-v22.1 TODOs

This commit fixes up any post-v22.1 TODOs with unimplemented messages
or comments annotations for the future.

Release note (sql change): We introduce a new row level TTL feature to
CockroachDB, which is available as a beta feature. This allows users to
use a special syntax to automatically mark rows for deletion. Rows are
deleted using a SCHEDULED JOB.

A user can create a table with TTL using:
```
CREATE TABLE t (id INT PRIMARY KEY) WITH (ttl_expire_after = '10 mins')
```

Where `ttl_expire_after` is a duration expression. A user can also add
TTL to an existing table using
```
ALTER TABLE t SET (ttl_expire_after = '10 mins')
```

This creates a new column, `crdb_internal_expiration`, which
automatically is set to `now() + ttl_expire_after` when inserted by
default or on update. The scheduled job will delete any rows which
exceed this timestamp as of the beginning of the job run.

The TTL job is configurable in a few ways using the WITH/SET syntax:
* ttl_select_batch_size: how many rows to select at once (by default
  it is cluster setting sql.ttl.default_select_batch_size)
* ttl_delete_batch_size: how many rows to delete at once (default
  cluster setting sql.ttl.default_select_batch_size)
* ttl_delete_rate_limit: maximum rows to delete per second for
  the given table (default cluster setting
  sql.default.default_delete_rate_limit)
* ttl_pause: pauses the TTL job (also globally pausable with
  `sql.ttl.job.enabled`).

Using `ALTER TABLE table_name RESET (<parameter>)` will reset the
parameter to re-use the default, or `RESET(ttl)` will
disable the TTL job for the table and remove the
`crdb_internal_expiration` column.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76918)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/catalog/tabledesc/structured.go, line 2527 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76918#-MwlU0TK9dfoRuv7Artf:-MwmWo908XpZ6SOWjacV:b-omz0ki) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b356c43de90fa473be4c43d1e81e67880afd178/pkg/sql/catalog/tabledesc/structured.go#L2527)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i wonder if there's a good way to make sure we don't forget to update this function when new storage params are added
</blockquote></details>

maybe check all CREATE TABLE in logic test is self reproducing, heh....


<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss
